### ISSUE TITLE
Partition and Row Key mapping to corresponding properties does not work with KeySeparator value. It always splits the properties by the first underscore

### ISSUE THREAD

**Issue by ajoylabroo** [s-innovations/azure-table-storage-repository-pattern#4]
Hello, I have one more issue for ya!!

In an example of [UnitTest5.cs](https://github.com/s-innovations/azure-table-storage-repository-pattern/blob/master/S-Innovations.Azure.TableStorageRepository.Test/UnitTest5.cs)

If you have an underscore in the one of property values that form a composite Row key, for example:

``` csharp
...
protected override void OnModelCreating(TableStorageModelBuilder modelbuilder)
{
    modelbuilder.Entity<Product>()
    .HasKeys(pk => pk.Category, rk =>new {  rk.Store, rk.ProductId})
    //   .HasKeys(pk => pk.Category,rk=>string.Format("{0}_{1}", rk.Store.ToLower(), rk.ProductId))              
    .ToTable("unittest5");

    base.OnModelCreating(modelbuilder);
}
...
var cat = "test";
            var ctx = new ProductContext();
            ctx.Products.Add(new Product
            {
                ProductId = 1,
                Name = "MyName",
                ImageName = "atest",
                Category = cat,
                Description = "adsad",
                Store = "test_abc",
            });
            ctx.SaveChangesAsync().Wait();

            var products = (from f in ctx.Products
                            where f.Category == cat
                            select f).ToArray();
```

In the above example `Store = "test_abc"` and `ProductId = 1`, so now the RowKey becomes 
`test_abc__1` [because if KeySeparator property is not set, it uses __ (two underscores) as default]. 

However, after running the following statement,

```
var products = (from f in ctx.Products
                            where f.Category == cat
                            select f).ToArray();
```

this is the output I get

```
products[0].Store = "test"
products[0].ProductId = "abc__1"
```

instead of getting

```
products[0].Store = "test_abc"
products[0].ProductId = "1"
```

It somehow still uses the first underscore to de-construct the properties from a RowKey. It does that for PartitionKey as well. I have tried explicitly setting the `KeySeparator = "|"`, it still won't split the property on KeySeparator value but only on first underscore.

Hope you can help. Thanks.

**Comment by pksorensen** [comment]
Cool. thats clearly not working as it was intended :) I will try to resolve it in the weekend.

**Comment by ajoylabroo** [comment]
Thanks.

Sent via the Samsung Galaxy Note® 3

**Comment by ajoylabroo** [comment]
Hello,

I would like to take this opportunity to say thank you for this repository. It has definitely helped me speed up my development work. I just wanted to let you know that a part of my project is dependent on your fix and whenever you can get to it, would be great. Thanks.

**Comment by pksorensen** [comment]
Sorry about not getting back. Can you verify that the unit test fails? On my end the unittest5 do work as expected? 

Could you make sure that you are using a empty table "unittest5" ->"unittest6" and the problem is not caused with some bad data added prior?

**Comment by pksorensen** [comment]
Also, using a | in keys are not working according to :http://stackoverflow.com/questions/11514707/azure-table-storage-rowkey-restricted-character-patterns 

so `KeySeparator = "|"` properly is a no go.

**Comment by ajoylabroo** [comment]
Okay, but even if I use `KeySeparator = "__"`, it doesn't work.

**Comment by pksorensen** [comment]
Do the Unittest5 fails? It do not fail for me - so could you show the test out put or tell me how to change the unit test such it fails and reproduce your problem?

**Comment by ajoylabroo** [comment]
Absolutely, let me run UnitTest5 and I will write back the results soon.

**Comment by ajoylabroo** [comment]
This is weird, but my Test passes too. Although when I am using the same code in the actual project, it fails to correctly assign the components of PartitionKey to its corresponding properties. Let me check again, if there is something else going in. I will keep you posted. Thanks for looking into it.

**Comment by pksorensen** [comment]
I would assume that you somehow by mistake got added the wrong data to the table while developing in the beginning. Could you use a table storage tool to inspect the table and verify that you dont have some bad entries?

**Comment by ajoylabroo** [comment]
Hi Poul,

Things are working correctly now. There was some dummy data in the table that had PartitionKey and RowKey assigned using the old way, but it also contained the data using the new assignment `.HasKeys(pk => new { pk.Category, pk.StoreSanitized }, rk => new { rk.ProductId })` For some reason it didn't assign the properties on retrieval correctly for even the correct rows. After removing the incorrect rows from the table, it started working correctly now. Thanks for taking the time to respond and your suggestions.

**Comment by ajoylabroo** [comment]
Looks like the issue persists. It is weird in a way because it works well when I have a breakpoint set anywhere after `ProductContext` is instantiated and before the `select` linq statement. If I run the web page, which calls this function without a breakpoint, it fails to correctly split the Primary Key on ( __ ) . As if, there is fraction of a delay in getting that split to work. The code doesn't get time to properly split the key when run without the breakpoint. 

Test cases work all the time though. It would be awesome if you could replicate the error, but let me do some more digging and I will let you know what I find. Thanks.

**Comment by pksorensen** [comment]
I have two busy days today and tomorrow. I will try to replicate also from what you are describing - but before we can reproduce the error in a unit test - it will be hard to fix.
